File: hacker_rank/heap.cpp
#include<iostream>
#include<vector>

namespace {

		template< typename Data >
		void swap( std::vector< Data >& heap, int p, int r ){
			Data temp = heap[p];
			heap[p] = heap[r];
			heap[r] = temp;

		}
}

template <typename Data>
class Heap final{
	public:
		Heap(){
			m_size = 0;	
			m_heap.push_back( Data() );
		}

		Heap( const Heap& heap ){

		}

		Heap& operator=( const Heap& heap ){

			return *this;
		}

		void addElement( Data element ){
			m_heap.push_back( element );
			m_size++;
				
		}

		void deleteElement( Data element ){

		}

		void printMin(){

		}

	private:
		std::vector< Data > m_heap;
		int m_size;

		void minHeapify( int index ){
			int element_index = m_size-1
			int parent_index = element_index / 2;
			if( parent_index > 0 && m_heap[m_size] < m_heap[parent_index] ){
				swap<Data>( m_heap, parent_index, element_index );
				
			}

		}

};

File: hacker_rank/heap.py

def swap( heap, p, r ):
    temp = heap[p]
    heap[p] = heap[r]
    heap[r] = temp


def minHeapify( heap, index, p, r ):
    min_index = index
    if p < len( heap ) and heap[p] < heap[min_index]:
        min_index = p
    if r < len( heap ) and heap[r] < heap[min_index]:
        min_index = r

    if min_index != index:
        swap( heap, min_index, index )
        minHeapify( heap, min_index, p, r )


def buildHeap( heap ):
    for i in range(0, len( heap ) / 2 ):
        minHeapify( heap, i , 0, len( heap ) - 1 )

def addElement( heap, element ):
    heap.append( element )
    buildHeap( heap )

def deleteElement( heap, element ):
    index = 0
    for i in range( 0, len( heap ) ):
        if heap[i] == element:
            index = i
            break

    swap( heap, index, len( heap ) - 1 )
    minHeapify( heap, index, 0, len( heap ) - 1 )

def printMin( heap ):
    print heap[0]

import sys

heap = [] 

def getInput():
    line = sys.stdin.readline().strip( "\n" )




